enclave-runtime/src/shard_vault.rs

```diff
@@ -0,0 +1,205 @@
+/*
+	Copyright 2021 Integritee AG
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+use crate::{
+	error::{Error, Result as EnclaveResult},
+	initialization::global_components::{
+		GLOBAL_OCALL_API_COMPONENT, GLOBAL_SIGNING_KEY_REPOSITORY_COMPONENT,
+		GLOBAL_STATE_HANDLER_COMPONENT,
+	},
+	utils::{
+		get_extrinsic_factory_from_integritee_solo_or_parachain,
+		get_node_metadata_repository_from_integritee_solo_or_parachain,
+	},
+};
+use codec::{Compact, Decode, Encode};
+use itp_component_container::ComponentGetter;
+use itp_extrinsics_factory::CreateExtrinsics;
+use itp_node_api::{
+	api_client::{PairSignature, StaticExtrinsicSigner},
+	metadata::{
+		pallet_proxy::PROXY_DEPOSIT,
+		provider::{AccessNodeMetadata, Error as MetadataProviderError},
+	},
+};
+use itp_node_api_metadata::pallet_proxy::ProxyCallIndexes;
+use itp_nonce_cache::NonceCache;
+use itp_ocall_api::EnclaveOnChainOCallApi;
+use itp_sgx_crypto::key_repository::AccessKey;
+use itp_stf_interface::SHARD_VAULT_KEY;
+use itp_stf_state_handler::{handle_state::HandleState, query_shard_state::QueryShardState};
+use itp_types::{
+	parentchain::{AccountId, Address, ParentchainId, ProxyType},
+	OpaqueCall, ShardIdentifier,
+};
+use log::*;
+use sgx_types::sgx_status_t;
+use sp_core::crypto::{DeriveJunction, Pair};
+use std::{slice, sync::Arc, vec::Vec};
+
+#[no_mangle]
+pub unsafe extern "C" fn init_proxied_shard_vault(
+	shard: *const u8,
+	shard_size: u32,
+) -> sgx_status_t {
+	let shard_identifier =
+		ShardIdentifier::from_slice(slice::from_raw_parts(shard, shard_size as usize));
```

Comment:
I am so wary of our inconsistent handling here, sometimes we use `decode`, sometimes `from_slice`, I really hope we can fix this ffi handling somewhen in the future. Maybe even completely abstract it away.

Reply:
should be unified in another PR IMO

Reply:
Yes, but I wanted to raise awareness.